src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java

```diff
@@ -1,86 +1,212 @@
+  /**
+   * Factory method to create a new instance of {@link CommandContext} based on the schema object we
+   * are working with
+   *
+   * @param schemaObject
+   * @param embeddingProvider
+   * @param commandName
+   * @param jsonProcessingMetricsReporter
+   * @return
+   */
+  public static CommandContext<KeyspaceSchemaObject> forSchemaObject(
+      KeyspaceSchemaObject schemaObject,
+      EmbeddingProvider embeddingProvider,
+      String commandName,
+      JsonProcessingMetricsReporter jsonProcessingMetricsReporter) {
+    return new CommandContext<>(
+        schemaObject, embeddingProvider, commandName, jsonProcessingMetricsReporter);
+  }
 
   /**
-   * @return Returns empty command context, having both {@link #namespace} and {@link #collection}
-   *     as <code>null</code>.
+   * Factory method to create a new instance of {@link CommandContext} based on the schema object we
+   * are working with
+   *
+   * @param schemaObject
+   * @param embeddingProvider
+   * @param commandName
+   * @param jsonProcessingMetricsReporter
+   * @return
    */
-  public static CommandContext empty() {
-    return EMPTY;
+  public static CommandContext<DatabaseSchemaObject> forSchemaObject(
+      DatabaseSchemaObject schemaObject,
+      EmbeddingProvider embeddingProvider,
+      String commandName,
+      JsonProcessingMetricsReporter jsonProcessingMetricsReporter) {
+    return new CommandContext<>(
+        schemaObject, embeddingProvider, commandName, jsonProcessingMetricsReporter);
   }
 
-  public CollectionSettings.SimilarityFunction similarityFunction() {
-    return collectionSettings.vectorConfig().similarityFunction();
+  @SuppressWarnings("unchecked")
+  public CommandContext<CollectionSchemaObject> asCollectionContext() {
+    checkSchemaObjectType(CollectionSchemaObject.TYPE);
+    return (CommandContext<CollectionSchemaObject>) this;
   }
 
-  public boolean isVectorEnabled() {
-    return collectionSettings.vectorConfig() != null
-        && collectionSettings.vectorConfig().vectorEnabled();
+  @SuppressWarnings("unchecked")
+  public CommandContext<TableSchemaObject> asTableContext() {
+    checkSchemaObjectType(TableSchemaObject.TYPE);
+    return (CommandContext<TableSchemaObject>) this;
   }
 
-  public IndexingProjector indexingProjector() {
-    return collectionSettings.indexingProjector();
+  @SuppressWarnings("unchecked")
+  public CommandContext<KeyspaceSchemaObject> asKeyspaceContext() {
+    checkSchemaObjectType(KeyspaceSchemaObject.TYPE);
+    return (CommandContext<KeyspaceSchemaObject>) this;
   }
+
+  @SuppressWarnings("unchecked")
+  public CommandContext<DatabaseSchemaObject> asDatabaseContext() {
+    checkSchemaObjectType(DatabaseSchemaObject.TYPE);
+    return (CommandContext<DatabaseSchemaObject>) this;
+  }
+
+  private void checkSchemaObjectType(SchemaObject.SchemaObjectType expectedType) {
+    Preconditions.checkArgument(
```

Comment:
We can discuss whether we should start using Guava Preconditions or not -- just noting for now (new usage, not used before although we do use other parts of Guava already)

Reply:
Postponing discussions.